File: src/main/kotlin/com/nylas/BuildInfo.java
package com.nylas;

import java.io.FileInputStream;
import java.io.InputStream;
import java.util.Properties;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class BuildInfo {
	
	private static final Logger log = LoggerFactory.getLogger(BuildInfo.class);

	static final String VERSION;
	static final String COMMIT_ID;
	static final String TIMESTAMP;
	
	static {
		
		String version = null;
		String commitId = null;
		String timestamp = null;
		try {
			// First, look for build.properties file left by gradle
			try(InputStream in = BuildInfo.class.getResourceAsStream("/build.properties")) {
				if (in != null) {
					Properties props = new Properties();
					props.load(in);
					version = props.getProperty("version");
					commitId = props.getProperty("commit.hash");
					timestamp = props.getProperty("build.timestamp");
				}
			}
		
			// Otherwise, look directly for gradle.properties file (e.g. in an IDE)
			if (version == null || version.isEmpty()) {
				try (InputStream in = new FileInputStream("gradle.properties")) {
					if (in != null) {
						Properties props = new Properties();
						props.load(in);
						version = props.getProperty("version");
					}
				}
			}
		
		} catch (Throwable t) {
			log.error("Failed to load build info", t);
		}

		VERSION = version;
		COMMIT_ID = commitId;
		TIMESTAMP = timestamp;
	}
}

File: src/main/kotlin/com/nylas/BuildInfo.kt
package com.nylas

import org.slf4j.LoggerFactory
import java.io.FileInputStream
import java.util.*

/**
 * @suppress Not for public use.
 */
object BuildInfo {
  private val log = LoggerFactory.getLogger(BuildInfo::class.java)
  var VERSION: String? = null

  init {
    try {
      // First, look for build.properties file left by gradle
      BuildInfo::class.java.getResourceAsStream("/build.properties").use { `in` ->
        if (`in` != null) {
          val props = Properties()
          props.load(`in`)
          VERSION = props.getProperty("nylas.build.version")
        }
      }

      // Otherwise, look directly for gradle.properties file (e.g. in an IDE)
      if (VERSION == null || VERSION!!.isEmpty()) {
        FileInputStream("gradle.properties").use { `in` ->
          val props = Properties()
          props.load(`in`)
          VERSION = props.getProperty("version")
        }
      }
    } catch (t: Throwable) {
      log.error("Failed to load build info", t)
    }
  }
}
